README
Initial Blog annoucement:

  http://ifdeflinux.blogspot.com/2012/10/procenv-and-process-environment.html

Homepage:

  https://launchpad.net/procenv

This is a simple utility, licensed under the GPL, that essentially dumps
every conceivable aspect of a processes environment (*) that it can
(without needing super-user privs). It unashamedly emulates a number of
existing system utilities as it is attempting to be all-encompassing: I
wrote it with the aim of being able to dump "everything" that a process
may care about by simply running a single program (by default). Also,
the line of demarcation between "process", "program" and  "system" is
slightly blurry in some aspects. For example sysconf(3) variables could
arguably be considered system attributes, but procenv shows these too
since they are obviously meant to be querable by applications.


(*) - If you find anything missing, please either raise a bug via
      https://bugs.launchpad.net/procenv/+filebug or send a patch! :-)


NEWS
0.60	2021-07-13

	* Bump output format version (should have been done for 0.58).
	* snap: Add missing staging packages.
	* autoconf: Fix AppArmor and SELinux detection
	  Bug fix that necessitated a new release: previously, the security
	  context displayed with `--misc` could show as "unknown" due to a
	  bug in `configure.ac`'s detection code.
	* docs: Add snap details to README.

0.59	2021-07-13

	* doc improvements (README and man page).
	* snap: Add snapcraft config file
	* `--sysconf`:
	  - Display output in order.
	  - Removed duplicate entry for `_SC_EXPR_NEST_MAX`.
	  - Added new entries for:
	    - `_SC_2_PBS_CHECKPOINT`
	    - `_SC_SS_REPL_MAX`
	    - `_SC_STREAMS`
	    - `_SC_TRACE_EVENT_NAME_MAX`
	    - `_SC_TRACE_NAME_MAX`
	    - `_SC_TRACE_SYS_MAX`
	    - `_SC_TRACE_USER_EVENT_MAX`
	    - `_SC_XOPEN_STREAMS`
	* `--misc`: BUG fix: Check for `PR_GET_UNALIGN` `prctl`, **not**
	  `PR_GET_UNALIGNED`.
	* `--namespaces`: BUG fix: Don't fully resolve namespace links to handle
	  new `pid_for_children` and `time_for_children` namespaces.

0.58	2021-06-25

	* darwin: Support `--libs`.
	* darwin: Support `AF_LINK` for `--network`.
	* docs: Remove TOC.

0.57	2021-06-19

	* Add macports installation instructions (thanks Haren).
	* --timezone: Enabled for Darwin, FreeBSD, NetBSD and OpenBSD.
	* --clocks: Add `CLOCK_BOOTTIME_ALARM`, `CLOCK_PROCESS_CPUTIME_ID`
	  and `CLOCK_REALTIME_ALARM` to output.
	* darwin: Enabled `--clocks`.

0.56	2021-06-13

	* tidy: compact and align structs for memory size and performance.
	* darwin: Add memory details (total, free, wired, unused, active and
	  inactive memory).
	* doc improvements.

0.55	2021-04-08

	* Documentation improvements.
	* Hurd build fixes:
	  - Disable detailed memory reporting as `sysinfo()` not available).
	  - Fixed detection for modern versions of Hurd.
	* Nits.

0.54	2021-03-15

	* --capabilities: Added new capabilities:
	  - `CAP_BPF`
	  - `CAP_CHECKPOINT_RESTORE`
	  - `CAP_PERFMON`
	* --clocks: Add CLOCK_TAI.
	* --memory: Add details of total and available memory and swap.
	  Requires the `libsysinfo` package on *BSD.
	* --process: Fixed nasty bug (infinite loop) on *BSD if any of
	  the following sysconf's are set:
	  - `security.bsd.see_other_uids=0`
	  - `security.bsd.see_other_gids=0`
	  - `kern.randompid=1`
	* Fix compiler flag checking in the configure script.

0.53	2021-03-01

	* Really fix FreeBSD capabilities/capsicum build.

0.52	2021-02-28

	* Fix GCC-10 build when used with `-Werror=format-overflow`
	  (thanks Lukas Maerdian).
	* Fix FreeBSD capabilities/capsicum build (thanks Li-Wen Hsu).

0.51	2019-08-02

	* Fixed failure scenario identified by `scan-build(1)`.
	* Fixed GNU Hurd detection.
	* Fixed building without `libcap` (thanks Dave Love).
	* Fix compiler warnings for gcc9 (thanks Vorlon).

0.50	2017-10-16

	* --compiler: Add additional feature test macros.
	* --misc: Show if running in a virtual machine.

0.49	2017-02-12

	* FreeBSD 11 capsicum fixes.
	* Fix PROCENV_EXEC_ENV harder.

0.48	2017-01-31

	* Fix makedev(3) compiler error (it's now in sys/sysmacros.h).
	  (thanks Dave Love).

0.47	2017-01-31

	* Fixed hang when using PROCENV_EXEC variable.
	* Test improvements.

0.46	2016-05-28

	* Fixed --file-append behaviour.
	* Test improvements.
	* Added support for Apple OSX (darwin).

0.45	2016-03-21

	* Code restructured internally to simplify maintenance and adding
	  new platforms. See src/platform/README.rst (and README-BUILD.rst).

	* Dropped support for:
	  - kFreeBSD (Debian using a FreeBSD kernel) - defunc project.
	  - Ubuntu Lucid (10.04) - no longer a supported platform.

	* Added support for NetBSD, OpenBSD and Minix 3.

	* --meta: Now shows details of the procenv platform driver being
	  used.

	* --misc now shows "security module" rather than
	  "linux security module".

	* --file-append: new option (env var is "PROCENV_FILE_APPEND")
	  to append to the --file= specified rather than overwriting.

0.44	2016-02-15

	* --cgroup: Handle cgroup2 (cgroups v2).

	* --clocks: now also includes CLOCK_REALTIME_COURSE, CLOCK_REALTIME_HR,
	  CLOCK_MONOTONIC_COURSE, CLOCK_MONOTONIC_RAW and CLOCK_BOOTTIME.

	* --cpu: added I/O priority.

	* Fixed crash if non-output argument specified before "--exec".

0.43	2015-11-24

	* Namespace fix for non-Linux platforms.

0.42	2015-11-05

	* Documentation improvements.
	* Removed reliance on autoreconf
	  (required due to limitations in github release process).
	* Improved '--namespaces' output by removing indexes.

0.41	2015-10-23

	* Further test improvements.
	* Correct output order (libc before misc).
	* Run compiler test before building procenv to aide remote debugging
	  should the build fail.
	* Code tidy-ups and Coverity tweaks.
	* Added '--namespaces'/'-F' option.

0.40	2015-09-27

	* Updated email address.
	* New --libc'/'-B' option.
	* --enable-reproducible-build
	* '--compiler': Added _DEFAULT_SOURCE, _LARGEFILE_SUPPORT and
	  __STDC_VERSION__.
	* locale fixes.

0.39	2015-08-25

	* Fixed tests harder :)

0.38	2015-08-24

	* Fixed tests for non-UTF-8 locales.

0.37	2015-08-23

	* Improved error handling.
	* Added CAP_AUDIT_READ capability.
	* Improved build-time tests.
	* Tolerate a completely empty environment.
	* Fixed bug where fsid in '--mounts' output was displaying truncated
	  output on Linux.
	* Fix crasher bug triggered by specifying a UTF-8 character to
	  '--separator' or '--crumb-separator'.
	* Allow '--indent-char' to be UTF-8.

0.36	2014-08-16

	* RHEL/Centos spec file - thanks Dave Love
	  (LP: #1327594).
	* Improved checks for SELinux/Apparmor (LP: #1333182).
	* Added missing types from stdint.h to '--sizeof'.
	* Fixed building on systems with old NUMA libraries, such as
	  Centos/RHEL 5 (LP: #1333194). However, bug is not yet fully
	  resolved since cpu details seem to be reported incorrectly
	  on these old systems.
	* Corrected SELinux detection code and added MLS detection.

src/procenv.h
/*--------------------------------------------------------------------
 * Copyright (c) 2012-2021 James O. D. Hunt <jamesodhunt@gmail.com>.
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 *--------------------------------------------------------------------
 */

#ifndef PROCENV_H
#define PROCENV_H

/* for dl_iterate_phdr(3) */
#ifndef _GNU_SOURCE
#define _GNU_SOURCE
#endif

#ifdef HAVE_CONFIG_H
# include <config.h>
#endif /* HAVE_CONFIG_H */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <syslog.h>
#include <stdarg.h>
#include <stddef.h>
#include <stdint.h>
#include <stdbool.h>
#include <float.h>
#include <wchar.h>
#include <wctype.h>
#include <unistd.h>
#include <limits.h>
#include <libgen.h>
#include <time.h>
#include <signal.h>
#include <ctype.h>
#include <errno.h>
#include <paths.h>
#include <dirent.h>
#include <pwd.h>
#include <grp.h>

/* XXX: must come before sched.h on Minix! */
#include <pthread.h>

#include <sched.h>
#include <termios.h>
#include <getopt.h>
#include <assert.h>
#include <sys/types.h>
#include <inttypes.h>
#include <sys/ioctl.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <fenv.h>
#include <sys/utsname.h>
#include <locale.h>
#include <sys/socket.h>
#include <netdb.h>
#include <ifaddrs.h>
#include <arpa/inet.h>
#include <sys/ipc.h>
#include <sys/shm.h>
#include <sys/sem.h>
#include <sys/msg.h>
#include <sys/param.h>
#include <sys/time.h>
#include <sys/resource.h>

#if !defined (PROCENV_PLATFORM_DARWIN)
#include <sys/sysinfo.h>
#endif

#include "util.h"
#include "string-util.h"
#include "pr_list.h"
#include "output.h"
#include "platform.h"

#include "platform-headers.h"

/*********************************************************************/
// FIXME: ANDROID - need to auto-generate values!!

#if defined PROCENV_PLATFORM_ANDROID

#ifndef PACKAGE_NAME
#define PACKAGE_NAME "procenv"
#endif

#ifndef PACKAGE_VERSION
#define PACKAGE_VERSION "0.24"
#endif

#ifndef PACKAGE_STRING
#define PACKAGE_STRING PACKAGE_NAME
#endif

/* major(3) / minor(3) */
#include <sys/sysmacros.h>

#endif /* PROCENV_PLATFORM_ANDROID */

/*********************************************************************/

/**
 * PROCENV_FORMAT_VERSION:
 *
 * Version of output format.
 *
 * XXX: must be updated for every change.
 *
 * VERSION 1:
 *   - Original format (up to procenv 0.27).
 * VERSION 2:
 *   - Added --memory.
 *   - Expanded --cpu.
 *   - Added missing --cpu to default output.
 *   - Moved memory page size from --misc to --memory.
 * VERSION 3:
 *   - --fds: Added Capsicum capabilities for FreeBSD (alas this version
 *     is a NOP for non-BSD platforms).
 * VERSION 4:
 *   - Added "symbolic" values to --range output.
 * VERSION 5:
 *   - Capabilities output changed on linux to show not just the
 *     bounding set, but also permitted, inheritable and enabled values
 *     along with the numeric value of the define.
 * VERSION 6:
 *   - --misc: Added personality.
 * VERSION 7:
 *   - --confstr and pathconf values in --mount now show
 *     NA_STR rather than UNKNOWN_STR. --sysconf now shows NA_STR rather
 *     than -1.
 * VERSION 8:
 *   - --memory: Added NUMA API version.
 *   - --shared-memory: Added swap_attempts and swap_successes.
 * VERSION 9:
 *   - --meta now includes a 'build-type' field.
 * VERSION 10:
 *   - Output is now sorted correctly (again - previously show_misc()
 *     and show_libc() were being called in the wrong order).
 * VERSION 11:
 *   - Added --namespaces.
 * VERSION 12:
 *   - Change format of --namespaces output.
 * VERSION 13:
 *   - Added I/O priority to --cpu.
 * VERSION 14:
 *   - Added driver to --meta.
 * VERSION 15:
 *   - Renamed "linux security module" to "security" module in --misc.
 *   - Added procenv driver details to --meta.
 * VERSION 16:
 *   - Sort feature-test macros.
 * VERSION 17:
 *   - Added 'vm' to --misc output.
 * VERSION 18:
 *   - Added 'ambient' to --capabilities output.
 * VERSION 19:
 *  - Added more details to --memory and new capabilities to --capabilities.
 * VERSION 20:
 *  - More --memory details for Darwin.
 * VERSION 21:
 *  - Added `CLOCK_BOOTTIME_ALARM`, `CLOCK_PROCESS_CPUTIME_ID`
 *    and `CLOCK_REALTIME_ALARM` to --clocks output and supported --timezone
 *    on Darwin.
 * VERSION 22:
 *  - Supported --libs on Darwin.
 * VERSION 23:
 *  - Changed output order of --sysconf and --confstr, added various new
 *    sysconf values, plus changed output format of --namespaces.
 **/
#define PROCENV_FORMAT_VERSION 23

#if defined (PROCENV_PLATFORM_LINUX) || defined (PROCENV_PLATFORM_HURD)

#if ! defined (_LINUX_CAPABILITY_VERSION_3) && ! defined (CAP_LAST_CAP)
/* Ugh */
#define CAP_LAST_CAP 30
#endif

#endif /* PROCENV_PLATFORM_LINUX || PROCENV_PLATFORM_HURD */

/* Environment Variables */
#define PROCENV_OUTPUT_ENV           "PROCENV_OUTPUT"
#define PROCENV_FORMAT_ENV           "PROCENV_FORMAT"
#define PROCENV_FILE_ENV             "PROCENV_FILE"
#define PROCENV_FILE_APPEND_ENV      "PROCENV_FILE_APPEND"
#define PROCENV_EXEC_ENV             "PROCENV_EXEC"
#define PROCENV_INDENT_ENV           "PROCENV_INDENT"
#define PROCENV_INDENT_CHAR_ENV      "PROCENV_INDENT_CHAR"
#define PROCENV_SEPARATOR_ENV        "PROCENV_SEPARATOR"
#define PROCENV_CRUMB_SEPARATOR_ENV  "PROCENV_CRUMB_SEPARATOR"

#define PROCENV_BUFFER                1024

/* FIXME: explain! */
#define CTIME_BUFFER                  32

#include "messages.h"

/* Size of blocks we will show the user (as df(1) does) */
#define DF_BLOCK_SIZE 1024

#define PROGRAM_AUTHORS "James O. D. Hunt <jamesodhunt@gmail.com>"

#define type_hex_width(type) \
	(sizeof (type) * 2)

#define show_clock_res(clock) \
{ \
	struct timespec res; \
	section_open (#clock); \
	if (clock_getres (clock, &res) < 0) \
		entry ("resolution", "%s", UNKNOWN_STR); \
	else \
		entry ("resolution", "%ld.%09lds", res.tv_sec, res.tv_nsec); \
	section_close (); \
}

#define show_const(t, flag, constant) \
	object_open (false); \
    	entry (#constant, "%d", !!(t.flag & constant)); \
	object_close (false)

/**
 * Show a terminal special characters attribute.
 *
 * t: struct termios,
 * elem: element of c_cc array,
 * lock_status: struct termios representing lock status of @t.
 **/
#define show_cc_tty(t, elem, lock_status) \
    entry (#elem, "0x%x%s", \
	t.c_cc[elem], \
	lock_status.c_cc[elem] ? " (locked)" : "");

/**
 * Show a terminal attribute constant value.
 *
 * t: struct termios,
 * flag: name of attribute,
 * constant: value of @flag,
 * lock_status: struct termios representing lock status of @t.
 **/
#define show_const_tty(t, flag, constant, lock_status) \
	entry (#constant, \
		"%d%s", \
		!!(t.flag & constant), \
		!!(lock_status.flag) ? " (locked)" : "")

#define show_pathconf(what, path, name) \
{ \
	long conf; \
	errno = 0; \
	conf = pathconf (path, name); \
	if (conf == -1 && errno == 0) { \
	    entry (#name, "%s", NA_STR); \
	} else { \
	    entry (#name, "%d", conf); \
	} \
}

#define SPEED(s) \
	{#s, s}

#define mk_map_entry(s) \
	{#s, s}

#define show_confstr(s) \
{ \
    _show_confstr (s, #s); \
}

#define _show_confstr(s, name) \
{ \
	size_t len; \
	char *buffer; \
	\
	errno = 0; \
	len = confstr(s, NULL, 0); \
	if (len && errno == 0) { \
		\
		buffer = calloc (1, len); \
		if (! buffer) { \
			die ("failed to allocate space for confstr"); \
		} \
		\
		assert (confstr (s, buffer, len) == len); \
		\
		/* Convert multi-line values to multi-field */ \
		for (size_t i = 0; i < len; i++) { \
			if (buffer[i] == '\n') buffer[i] = ' '; \
		} \
		\
		entry (name, "%s%s%s", \
				buffer && buffer[0] ? "'" : "", \
				buffer && buffer[0] ? buffer : NA_STR, \
				buffer && buffer[0] ? "'" : ""); \
		\
		free (buffer); \
	} \
}

/* Note: param is ignored */
#define limit_max(l) \
	((unsigned long int)-1)

#define show_limit(limit) \
{ \
	struct rlimit tmp; \
	\
	if (getrlimit (limit, &tmp) < 0) { \
		die ("failed to query rlimit '%s'", #limit); \
	} \
	\
	section_open (#limit); \
	\
	section_open ("soft"); \
	entry ("current", "%lu", (unsigned long int)tmp.rlim_cur); \
	entry ("max", "%lu", limit_max (limit)); \
	section_close (); \
	\
	section_open ("hard"); \
	entry ("current", "%lu", (unsigned long int)tmp.rlim_max); \
	entry ("max", "%lu", limit_max (limit)); \
	section_close (); \
	\
	section_close (); \
}

#define show_usage(rusage, name) \
	entry (#name, "%lu", rusage.name)

#define get_sysconf(s) \
 	sysconf (s)

#define mk_sysconf_map_entry(name) \
	{#name, name }

#define show_sizeof_type(type) \
	entry (#type, "%lu byte%s", \
			(unsigned long int)sizeof (type), \
			sizeof (type) == 1 ? "" : "s")


#define show_size(type) \
	entry ("size", "%lu byte%s", \
			(unsigned long int)sizeof (type), \
			sizeof (type) == 1 ? "" : "s")

#define free_if_set(ptr) \
    if (ptr) free (ptr)

char *get_path (const char *argv0);
char *get_personality_flags (unsigned int flags);
const char *get_arch (void);
const char *get_group_name (gid_t gid);
const char *get_os (void);
const char *get_personality_name (unsigned int domain);
const char *get_signal_name (int signum);
const char *get_thread_scheduler_name (int sched);
const char *get_user_name (gid_t gid);
int get_major_minor (const char *path, unsigned int *_major, unsigned int *_minor);
long get_kernel_bits (void);
void get_priorities (void);
void get_uname (void);

void show_all_groups (void);
void show_capabilities (void);
void show_cgroups (void);
void show_clocks (void);
void show_compiler (void);
void show_confstrs (void);
void show_data_model (void);
void show_libc (void);
void show_libs (void);
void show_locale (void);
void show_mounts (ShowMountType what);
void show_misc (void);
void show_msg_queues (void);
void show_network (void);
void show_oom (void);
void show_ranges (void);
void show_rlimits (void);
void show_semaphores (void);
void show_shared_mem (void);
void show_sizeof (void);
void show_threads (void);
void show_time (void);
void show_timezone (void);
void show_tty_attrs (void);
void show_uname (void);
void show_version (void);

void handle_proc_branch (void);

char *format_perms (mode_t mode);
void format_time (const time_t *t, char *buffer, size_t len);
void restore_locale (void);
int qsort_compar (const void *a, const void *b);

extern char **environ;

extern Output output;
extern pstring *doc;

#endif /* PROCENV_H */
